### PR TITLE
Feature minteramm openmarkets

### PR DESCRIPTION
**Pull request by br4ve-nao**
For deleting elements from storage we have two options

(i)   Remove(index) : Resets value at index to default value. Array size doesn't change. 
(ii)  Pop() : Removes the last element. Size reduces by 1

Option (i) This doesn't reduce size even after removal so doesn't change much from what is already there.

Option (ii)[implementation] : Here the order of elements will get changed. We can store the element at last index to expired index and pop the last element effectively deleting the expired element. But in few scenarios we might not be able to clear all expired markets. See below 2 scenarios

```
~ marked items are expired items
Scenario 1: All expired items are removed
   ~1      2        ~3        4    5  6   
    6       2       ~3        4    5        
    6       2        5        4	      

Scenario 2: All expired items are not removed
   ~1~    ~2~     ~3~    ~4~       (0)  
   ~4~   ~2~      ~3~              (1)
   ~4~   ~3~                       (2) (loop ends: i> openMarkets.length)

```

One way to remove all expired markets is to 
capture indexes-> delete > shift array ( repeat this till. there are no more expired indexes to capture). But this will result in quadratic time complexity

### CODE REVIEW

**Review by zareth-san — CHANGES_REQUESTED** [review]
Hey br4ve-nao, thanks for contributing to the project! This is a great start, and improves my own approach. I like how you solved the "Stack Too Deep" error with making an internal function. I think that's a good tradeoff between readability and gas consumption, and I wouldn't be surprised if `solc` does some clever inlining anyway.

There are some changes that still need to be done though. Please see my inline comments. The most important one is the one about moving the expired market eviction logic into `claimAllExpiredTokens`, because it's in that function where we already check for expired markets, so it makes sense to do the check there.

Also, a couple more things:

1: Instead of merging this PR onto `sirenmarkets:master`, can you change the eventual merge branch to `sirenmarkets:33-remove-expired-markets`? I find it easier to make multiple small PRs onto a feature branch, then rebase them all into a single commit which we then merge into `master`. I suspect you'll have a PR for the implementation (i.e. this PR), a PR for updating the subgraph and any contract changes you need to make to accommodate that, and a PR for tests.

2. This implementation logic is fine if we just consider the contract layer, but it's not going to be backwards compatible with the subgraph, and thus it's going to break the webapp. To understand how, look at https://github.com/sirenmarkets/core/blob/master/subgraph/schema.graphql#L9 and https://github.com/sirenmarkets/core/blob/master/subgraph/src/mappings/market.ts#L54. In the schema.graphql file we give each `Market` a `marketIndex`, which is equal to its index in the `MarketsRegistry.marketsByAssets[assetPair]` array. That value gets set on the subgraph in the mappings/market.ts#L54 file. Prior to your PR, this logic worked fine because the `marketsByAssets` array was append-only, and thus the indexes of the Markets would never change. But now that we're evicting expired Markets and changing indexes, we need to reflect that in the subgraph. We could do this by adding an event `MarketEvicted` which includes the Market address, the AMM address, and the index being evicted, and then add a subgraph mapping to update the subgraph's `Market.marketIndex` correctly. 

3. Needs tests!

Again, thank you for this contribution, and let me know if I can explain anything in more detail.

File: contracts/amm/MinterAmm.sol [hunk 1]
```diff
@@ -707,16 +713,20 @@ contract MinterAmm is InitializeableAmm, OwnableUpgradeSafe, Proxiable {
                     .balanceOf(address(optionMarket))
                     .mul(wTokenBalance)
                     .div(wTokenSupply));
+
+            //Store the element at last index to expired index and pop the last element.
+            openMarkets[i] = openMarkets[openMarkets.length-1];
```
[the remark this answers is not in the document]
Comment (br4ve-nao) [reply contracts/amm/MinterAmm.sol:718]: There is actually a if condition for expired markets in which i have enclosed the logic to delete the expired markets.

 if (optionMarket.state() == IMarket.MarketState.EXPIRED) {
...
..
openMarkets[i] = openMarkets[openMarkets.length-1];

}

File: contracts/amm/MinterAmm.sol [hunk 2]
```diff
@@ -676,16 +680,18 @@ contract MinterAmm is InitializeableAmm, OwnableUpgradeSafe, Proxiable {
...
-     * Get unclaimed collateral and payment tokens locked in expired wTokens
+     * Get unclaimed collateral and payment tokens locked in expired wTokens.
+     * and remove the expired markets from open markets.
+     * @dev : Removing expired markets use an approach to store the last element's value to expired index 
+     * and pop the last element. At times this might not clear all the expired markets
      */
-    function getUnclaimedBalances() public view returns (uint256, uint256) {
-        address[] memory markets = getMarkets();
-
+    function getUnclaimedBalances() public returns (uint256, uint256) {
```
[the remark this answers is not in the document]
Comment (br4ve-nao) [reply contracts/amm/MinterAmm.sol:688]: My Bad. I see now this was the function name that was mentioned in the issue that you had raised. Will make the changes.

File: contracts/amm/MinterAmm.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (br4ve-nao) [reply contracts/amm/MinterAmm.sol:718]: ok sure.

[the remark this answers is not in the document]
Comment (zareth-san) [reply contracts/amm/MinterAmm.sol:718]: ah gotcha, please indent this section of code so that it's clear it's inside of the `if` check

**Comment by zareth-san** [comment]
> For deleting elements from storage we have two options
> 
> (i) Remove(index) : Resets value at index to default value. Array size doesn't change.
> (ii) Pop() : Removes the last element. Size reduces by 1
> 
> Option (i) This doesn't reduce size even after removal so doesn't change much from what is already there.
> 
> Option (ii)[implementation] : Here the order of elements will get changed. We can store the element at last index to expired index and pop the last element effectively deleting the expired element. But in few scenarios we might not be able to clear all expired markets. See below 2 scenarios
> 
> ```
> ~ marked items are expired items
> Scenario 1: All expired items are removed
>    ~1      2        ~3        4    5  6   
>     6       2       ~3        4    5        
>     6       2        5        4	      
> 
> Scenario 2: All expired items are not removed
>    ~1~    ~2~     ~3~    ~4~       (0)  
>    ~4~   ~2~      ~3~              (1)
>    ~4~   ~3~                       (2) (loop ends: i> openMarkets.length)
> ```
> 
> One way to remove all expired markets is to
> capture indexes-> delete > shift array ( repeat this till. there are no more expired indexes to capture). But this will result in quadratic time complexity

I thought more about the problem you've stated here, namely that the naive approach of popping the last element and moving it to the i'th index doesn't work in certain edge cases. The other approach, where you need to shift all indexes, is also too gas expensive.

I think we could solve the problem in the naive approach by checking to see if the new element at the i'th index is itself EXPIRED, and if swap it with the last element and pop the now guaranteed-to-be-expired last element from the array. We keep doing this until the element at index i is NOT expired, or i == marketsArray.length. So something like this:


```solidity
for (let i = 0; i < openMarkets.length; i++) {
    IMarket optionMarket = IMarket(openMarkets[i]);
    // ... claim the expired market's wTokens, which means it can now be safely removed
    
    // remove expired markets
    while (optionMarket.state() == IMarket.EXPIRED && openMarkets.length != i) {
        openMarkets[i] = openMarkets[openMarkets.length-1];
        openMarkets.pop()

        // emit an event with the address of this AMM, and the index i
    }
}
```

I believe this works even though we're changing the length of the array as we're looping through it.

What do you think? See any potential edge-cases?

**Comment by br4ve-nao** [comment]
> > For deleting elements from storage we have two options
> > (i) Remove(index) : Resets value at index to default value. Array size doesn't change.
> > (ii) Pop() : Removes the last element. Size reduces by 1
> > Option (i) This doesn't reduce size even after removal so doesn't change much from what is already there.
> > Option (ii)[implementation] : Here the order of elements will get changed. We can store the element at last index to expired index and pop the last element effectively deleting the expired element. But in few scenarios we might not be able to clear all expired markets. See below 2 scenarios
> > ```
> > ~ marked items are expired items
> > Scenario 1: All expired items are removed
> >    ~1      2        ~3        4    5  6   
> >     6       2       ~3        4    5        
> >     6       2        5        4	      
> > 
> > Scenario 2: All expired items are not removed
> >    ~1~    ~2~     ~3~    ~4~       (0)  
> >    ~4~   ~2~      ~3~              (1)
> >    ~4~   ~3~                       (2) (loop ends: i> openMarkets.length)
> > ```
> > 
> > 
> > One way to remove all expired markets is to
> > capture indexes-> delete > shift array ( repeat this till. there are no more expired indexes to capture). But this will result in quadratic time complexity
> 
> I thought more about the problem you've stated here, namely that the naive approach of popping the last element and moving it to the i'th index doesn't work in certain edge cases. The other approach, where you need to shift all indexes, is also too gas expensive.
> 
> I think we could solve the problem in the naive approach by checking to see if the new element at the i'th index is itself EXPIRED, and if swap it with the last element and pop the now guaranteed-to-be-expired last element from the array. We keep doing this until the element at index i is NOT expired, or i == marketsArray.length. So something like this:
> 
> ```solidity
> for (let i = 0; i < openMarkets.length; i++) {
>     IMarket optionMarket = IMarket(openMarkets[i]);
>     // ... claim the expired market's wTokens, which means it can now be safely removed
>     
>     // remove expired markets
>     while (optionMarket.state() == IMarket.EXPIRED && openMarkets.length != i) {
>         openMarkets[i] = openMarkets[openMarkets.length-1];
>         openMarkets.pop()
> 
>         // emit an event with the address of this AMM, and the index i
>     }
> }
> ```
> 
> I believe this works even though we're changing the length of the array as we're looping through it.
> 
> What do you think? See any potential edge-cases?

I think this would work. Since we are always fetching the latest array length, I don't see a problem there. I tried a few examples and I don't see any potential edge-cases as of now but will think through to see if we have missed anything else.

**Comment by br4ve-nao** [comment]
> Hey br4ve-nao, thanks for contributing to the project! This is a great start, and improves my own approach. I like how you solved the "Stack Too Deep" error with making an internal function. I think that's a good tradeoff between readability and gas consumption, and I wouldn't be surprised if `solc` does some clever inlining anyway.
> 
> There are some changes that still need to be done though. Please see my inline comments. The most important one is the one about moving the expired market eviction logic into `claimAllExpiredTokens`, because it's in that function where we already check for expired markets, so it makes sense to do the check there.
> 
> Also, a couple more things:
> 
> 1: Instead of merging this PR onto `sirenmarkets:master`, can you change the eventual merge branch to `sirenmarkets:33-remove-expired-markets`? I find it easier to make multiple small PRs onto a feature branch, then rebase them all into a single commit which we then merge into `master`. I suspect you'll have a PR for the implementation (i.e. this PR), a PR for updating the subgraph and any contract changes you need to make to accommodate that, and a PR for tests.
> 
> 1. This implementation logic is fine if we just consider the contract layer, but it's not going to be backwards compatible with the subgraph, and thus it's going to break the webapp. To understand how, look at https://github.com/sirenmarkets/core/blob/master/subgraph/schema.graphql#L9 and https://github.com/sirenmarkets/core/blob/master/subgraph/src/mappings/market.ts#L54. In the schema.graphql file we give each `Market` a `marketIndex`, which is equal to its index in the `MarketsRegistry.marketsByAssets[assetPair]` array. That value gets set on the subgraph in the mappings/market.ts#L54 file. Prior to your PR, this logic worked fine because the `marketsByAssets` array was append-only, and thus the indexes of the Markets would never change. But now that we're evicting expired Markets and changing indexes, we need to reflect that in the subgraph. We could do this by adding an event `MarketEvicted` which includes the Market address, the AMM address, and the index being evicted, and then add a subgraph mapping to update the subgraph's `Market.marketIndex` correctly.
> 2. Needs tests!
> 
> Again, thank you for this contribution, and let me know if I can explain anything in more detail.

Had a few questions here:

-- If the index is used to access markets in marketsRegistry by the web-app, it would be fine since we are not removing from marketsRegistry.

-- If the index is used to access elements using getMarkets in minterAmm, then we need to pass in new address array in the event after the removal operation as the order of the elements in array would get changed.  If this is the case one way is we could pass in the address array in the marketEvicted event and sub-graph could capture and update indexes for all open markets. 

Also in the case of market that we removed I don't think we can remove an entry from sub-graph, we might have to mark as deleted using a flag and web app might have to handle accordingly

Let me know your thoughts on above.

Thanks

**Comment by br4ve-nao** [comment]
Moved expiry logic to getExpiredTokens().
Cleaned up code
Passes all tests
Only owner check on AddMarket in minteramm not present

File: test/amm/minterAmmRemoveExpired.js [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+
+const ONE_DAY = 86400
+const THIRTY_DAYS = 30 * ONE_DAY
+const THIRTY_TWO_DAYS = 32 * ONE_DAY
+const BTC_ORACLE_PRICE = 14_000 * 10 ** 8 // BTC oracle answer has 8 decimals places, same as BTC
+
+const STATE_OPEN = 0
+const STATE_EXPIRED = 1
+
+contract("Minter AMM Remove expired markets", (accounts) => {
```
Comment (zareth-san) [inline test/amm/minterAmmRemoveExpired.js:18]: Thanks for these tests; they cover  most of the edge cases I'm worried about.

Before merging, can you add additional checks in all 3 of the tests that verify the ending state of the `openMarkets` `address[]` is what we expect? So for instance for the first test you wrote, we should assert that `assert.equal(0, (await deployedAmm.getMarkets()).length)`, and for the second test we should see that:

we start with the ordering:
`[marketAddress1, marketAddress2, marketAddress3, marketAddress4, marketAddress5]` and then we end with `[marketAddress1, marketAddress2, marketAddress4]`

**Review by zareth-san — CHANGES_REQUESTED** [review]
This is looking almost ready to go! Please see my small inline comment about extending the tests, and once that's in I think this is good to merge (though I know there's still #36 and #37, as well as the subgraph changes, we'll deal with those later :)!)

**Comment by br4ve-nao** [comment]
Test cases have been updated with additional checks. Also I have included the test case that i had as part of #37 ,so you can ignore #37. The test case is to see whether the non-owner can addMarket to amm.

Overall following use-cases have been covered in the tests

-- Add markets to amm:
Non-owner shouldn't be allowed to add market to amm
owner should be allowed to add market to amm

-- All markets expired
All expired markets should be removed from the amm

-- 3 open & 2 expired markets
3 open should not be removed from amm
2 expired should be removed from the amm

-- 1 open & 1 markets expired(last one added to the array)
This is to handle the edge case where i becomes greater than the last index of the array. When you are
at the last position of index where market has expired and you remove it i will be
greater than last index and it could cause out of bounds exception

**Review by zareth-san — APPROVED** [review]
